pkg/datapath/loader/netlink.go

```diff
@@ -692,3 +695,62 @@ func renameDevice(from, to string) error {
 
 	return nil
 }
+
+// DeviceHasTCProgramLoaded checks whether a given device has tc filter/qdisc progs attached.
+func DeviceHasTCProgramLoaded(ifc string) (bool, error) {
+	const bpfProgPrefix = "cil_"
+
+	l, err := netlink.LinkByName(ifc)
+	if err != nil {
+		return false, fmt.Errorf("unable to find endpoint link by name: %w", err)
+	}
+
+	dd, err := netlink.QdiscList(l)
+	if err != nil {
+		return false, fmt.Errorf("unable to fetch qdisc list for endpoint: %w", err)
+	}
+	var found bool
+	for _, d := range dd {
+		if d.Type() == qdiscClsact {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return false, nil
+	}
+
+	ff, err := netlink.FilterList(l, netlink.HANDLE_MIN_INGRESS)
+	if err != nil {
+		return false, fmt.Errorf("unable to fetch ingress filter list: %w", err)
+	}
+	var filtersCount int
+	for _, f := range ff {
+		if filter, ok := f.(*netlink.BpfFilter); ok {
+			if strings.HasPrefix(filter.Name, bpfProgPrefix) {
+				filtersCount++
+			}
+		}
+	}
+	if filtersCount == 0 {
+		return false, nil
+	}
+
+	ff, err = netlink.FilterList(l, netlink.HANDLE_MIN_EGRESS)
+	if err != nil {
+		return false, fmt.Errorf("unable to fetch egress filter list: %w", err)
+	}
+	filtersCount = 0
+	for _, f := range ff {
+		if filter, ok := f.(*netlink.BpfFilter); ok {
+			if strings.HasPrefix(filter.Name, bpfProgPrefix) {
+				filtersCount++
+			}
+		}
+	}
+	if len(ff) > 0 && filtersCount == 0 {
```

Comment:
To me, it also looks like a more reliable check here would be based on `ep.RequireEgressProg()` rather than the presence of egress filters.

Reply:
Agreed!